### PR TITLE
DRV-90: Feature to enforce DateV and TimeV for encoder plus DateTimeExtensions

### PR DESCRIPTION
**Pull request by kcrandall**
Spoke to Marrony Neris on community slack about the issue with Time equal to zero always causing DateTime to be converted into DateV even if you want TimeV. This PR has changed he suggest in PR#90 plus other on slack

### CODE REVIEW

File: FaunaDB.Client/Types/Encoder.cs [hunk 1]
```diff
@@ -100,7 +100,7 @@ Value EncodeIntern(object obj)
                     return BooleanV.Of((bool)obj);
 
                 case TypeCode.DateTime:
-                    return Value.FromDateTime((DateTime)obj);
+                    return Value.FromDateTime((DateTime)obj, forceType);
```
Comment (marrony) [inline FaunaDB.Client/Types/Encoder.cs]: you need to pass `forceType` to `Value.FromDateTimeOffset()` below

File: FaunaDB.Client.Test/EncoderTest.cs [hunk 2]
```diff
@@ -325,5 +325,30 @@ public void TestEnumTypes()
             Assert.AreEqual(StringV.Of("ARM"), Encode(CpuTypes.ARM));
             Assert.AreEqual(StringV.Of("MIPS"), Encode(CpuTypes.MIPS));
         }
+        class DateTimeOverride
```
Comment (marrony) [inline FaunaDB.Client.Test/EncoderTest.cs]: would you mind add an empty line between the declarations

File: FaunaDB.Client.Test/EncoderTest.cs [hunk 3]
```diff
@@ -325,5 +325,30 @@ public void TestEnumTypes()
...
+            [FaunaDate]
+            public DateTime DateV { get; set; }
+
+            public DateTimeOverride(DateTime timeV, DateTime dateV)
+            {
+                this.TimeV = timeV;
+                this.DateV = dateV;
+            }
+        }
+        [Test]
```
Comment (marrony) [inline FaunaDB.Client.Test/EncoderTest.cs]: same here

File: FaunaDB.Client/Types/Attributes.cs [hunk 4]
```diff
@@ -49,13 +49,22 @@ public class FaunaFieldAttribute : Attribute
...
         {
             Name = name;
         }
     }
+    [AttributeUsage(AttributeTargets.Field | AttributeTargets.Property | AttributeTargets.Parameter)]
+    public class FaunaDate : Attribute
+    {
+
+    }
+    [AttributeUsage(AttributeTargets.Field | AttributeTargets.Property | AttributeTargets.Parameter)]
```
Comment (marrony) [inline FaunaDB.Client/Types/Attributes.cs]: empty line

File: FaunaDB.Client/Types/Attributes.cs [hunk 5]
```diff
@@ -49,13 +49,25 @@ public class FaunaFieldAttribute : Attribute
...
             Name = name;
         }
     }
 
+    [AttributeUsage(AttributeTargets.Field | AttributeTargets.Property | AttributeTargets.Parameter)]
+    public class FaunaDate : Attribute
+    {
+
+    }
+
```
Comment (evbruno) [inline FaunaDB.Client/Types/Attributes.cs]: Cold you please add some doc. and keep the braces in the same line
(see `FaunaIgnoreAttribute` or `FaunaConstructorAttribute`)

File: FaunaDB.Client/Types/Attributes.cs [hunk 6]
```diff
@@ -49,13 +49,25 @@ public class FaunaFieldAttribute : Attribute
...
+    [AttributeUsage(AttributeTargets.Field | AttributeTargets.Property | AttributeTargets.Parameter)]
+    public class FaunaTime : Attribute
+    {
+
+    }
+
     /// <summary>
```
Comment (evbruno) [inline FaunaDB.Client/Types/Attributes.cs]: same here :)

File: FaunaDB.Client/Types/Reflection.cs [hunk 7]
```diff
@@ -37,6 +37,22 @@ public static string GetName(this MemberInfo member)
...
         }
 
+        public static Type GetOverrideType(this MemberInfo member)
+        {
+            var date = member.GetCustomAttribute<FaunaDate>();
+            var time = member.GetCustomAttribute<FaunaTime>();
+
+            if (date != null && time != null)
+                throw (new InvalidOperationException("Can't use both FaunaDate and FaunaTime on the same property."));
+
```
Comment (evbruno) [inline FaunaDB.Client/Types/Reflection.cs]: you can also drop this "extra" parentheses around the `new` as well

File: FaunaDB.Client/Utils/DateTimeExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    public static class DateTimeExtensions
+    {
+        /// <summary>
+        /// Will return a Fauna <see cref="Types.TimeV"/> object.
+        /// </summary>
+        /// <returns> <see cref="Types.TimeV"/></returns>
+        public static Types.TimeV ToFaunaTime(this DateTime dt)
+        {
+            return new Types.TimeV(dt);
+        }
```
Comment (evbruno) [inline FaunaDB.Client/Utils/DateTimeExtensions.cs]: add a new line

File: FaunaDB.Client/Utils/DateTimeExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+        /// <summary>
+        /// Will return a Fauna <see cref="Types.TimeV"/> object.
+        /// </summary>
+        /// <returns> <see cref="Types.TimeV"/></returns>
+        public static Types.TimeV ToFaunaTime(this DateTimeOffset dt)
+        {
+            return new Types.TimeV(dt.UtcDateTime);
+        }
```
Comment (evbruno) [inline FaunaDB.Client/Utils/DateTimeExtensions.cs]: new line

File: FaunaDB.Client/Types/Encoder.cs [hunk 10]
```diff
@@ -46,7 +46,7 @@ class EncoderImpl
 
         Stack<object> stack = new Stack<object>();
 
-        public Value Encode(object obj)
+        public Value Encode(object obj, Type t = null)
```
Comment (evbruno) [inline FaunaDB.Client/Types/Encoder.cs]: rename `t` to `type`

**Comment by kcrandall** [comment]
Changed those files

**Comment by evbruno** [comment]
internal ticket: DRV-90

**Pull request merged**